File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseGameTeam.java
package com.yahyeet.boardbook.model.firebase.repository;

import com.google.firebase.firestore.DocumentSnapshot;
import com.yahyeet.boardbook.model.entity.Game;
import com.yahyeet.boardbook.model.entity.GameTeam;

import java.util.HashMap;
import java.util.Map;

class FirebaseGameTeam extends AbstractFirebaseEntity<GameTeam> {
	private String name;
	private String gameId;

	public FirebaseGameTeam() {
	}

	public FirebaseGameTeam(String id, String name, String gameId) {
		super(id);
		this.name = name;
		this.gameId = gameId;
	}

	@Override
	public Map<String, Object> toMap() {
		Map<String, Object> map = new HashMap<>();

		if (name != null) {
			map.put("name", name);
		}

		if (gameId != null) {
			map.put("gameId", gameId);
		}

		return map;
	}

	@Override
	public GameTeam toModelType() {
		GameTeam gameTeam = new GameTeam(name);
		gameTeam.setId(getId());
		gameTeam.setGame(new Game(gameId));
		return gameTeam;
	}

	public static FirebaseGameTeam fromModelType(GameTeam gameTeam) {
		return new FirebaseGameTeam(gameTeam.getId(), gameTeam.getName(), gameTeam.getGame().getId());
	}

	public static FirebaseGameTeam fromDocument(DocumentSnapshot document) {
		FirebaseGameTeam firebaseGameTeam = new FirebaseGameTeam();
		firebaseGameTeam.setId(document.getId());

		if (document.contains("name")) {
			firebaseGameTeam.setName(document.getString("name"));
		}

		if (document.contains("gameId")) {
			firebaseGameTeam.setGameId(document.getString("gameId"));
		}

		return firebaseGameTeam;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getGameId() {
		return gameId;
	}

	public void setGameId(String gameId) {
		this.gameId = gameId;
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseGameTeam.kt
package com.yahyeet.boardbook.model.firebase.repository

import com.google.firebase.firestore.DocumentSnapshot
import com.yahyeet.boardbook.model.entity.Game
import com.yahyeet.boardbook.model.entity.GameTeam

import java.util.HashMap

internal class FirebaseGameTeam : AbstractFirebaseEntity<GameTeam> {
    var name: String? = null
    var gameId: String? = null

    constructor() {}

    constructor(id: String, name: String, gameId: String) : super(id) {
        this.name = name
        this.gameId = gameId
    }

    override fun toMap(): Map<String, Any> {
        val map = HashMap<String, Any>()

        if (name != null) {
            map["name"] = name!!
        }

        if (gameId != null) {
            map["gameId"] = gameId!!
        }

        return map
    }

    override fun toModelType(): GameTeam {
        val gameTeam = GameTeam(name)
        gameTeam.id = id
        gameTeam.game = Game(gameId)
        return gameTeam
    }

    companion object {

        fun fromModelType(gameTeam: GameTeam): FirebaseGameTeam {
            return FirebaseGameTeam(gameTeam.id, gameTeam.name, gameTeam.game!!.id)
        }

        fun fromDocument(document: DocumentSnapshot): FirebaseGameTeam {
            val firebaseGameTeam = FirebaseGameTeam()
            firebaseGameTeam.id = document.id

            if (document.contains("name")) {
                firebaseGameTeam.name = document.getString("name")
            }

            if (document.contains("gameId")) {
                firebaseGameTeam.gameId = document.getString("gameId")
            }

            return firebaseGameTeam
        }
    }
}
